## Aula 03 - Filtragem Baseada em Conteúdo - Exercícios

In [3]:
import pandas as pd
import numpy as np

### Importar base de dados

In [4]:
!pip install wget
!python3 -m wget https://github.com/mmanzato/MBABigData/raw/master/ml-20m-compact.tar.gz
!tar -xvzf ml-20m-compact.tar.gz

  Preparing metadata (setup.py) ... done
  Created wheel for wget: filename=wget-3.2-py3-none-any.whl size=9655 sha256=a43131df48af3e1de59907c01f25c62425aa3a61226e6d47b700cc4ed540c850
  Stored in directory: /root/.cache/pip/wheels/8b/f1/7f/5c94f0a7a505ca1c81cd1d9208ae2064675d97582078e6c769
Successfully built wget

Saved under ml-20m-compact.tar.gz
dataset/
dataset/tags_sample.csv
dataset/._.DS_Store
dataset/.DS_Store
dataset/movies_sample.csv
dataset/._genome-tags.csv
dataset/genome-tags.csv
dataset/._ml-youtube.csv
dataset/ml-youtube.csv
dataset/._genome-scores.csv
dataset/genome-scores.csv
dataset/ratings_sample.csv


In [5]:
movies = pd.read_csv('./dataset/movies_sample.csv')
ratings = pd.read_csv('./dataset/ratings_sample.csv')
df = ratings[['userId', 'movieId', 'rating']]
df = df.merge(movies[['movieId', 'title']])
df

userId  movieId  rating                                     title
0           11     7481     5.0                         Enemy Mine (1985)
1          359     7481     4.0                         Enemy Mine (1985)
2          422     7481     3.0                         Enemy Mine (1985)
3          424     7481     3.0                         Enemy Mine (1985)
4          744     7481     3.0                         Enemy Mine (1985)
...        ...      ...     ...                                       ...
190616  131651   118606     3.5                               Dots (1940)
190617  134567   118606     3.5                               Dots (1940)
190618  131651    96257     1.0  Heroic Purgatory (Rengoku eroica) (1970)
190619  131651    98337     4.0                    97 Percent True (2008)
190620  131904   101666     2.0                 Night in Heaven, A (1983)

[190621 rows x 4 columns]

In [6]:
movies_tags = pd.read_csv('./dataset/tags_sample.csv')
movies_tags.head()

userId  movieId              tag  timestamp_y
0     279      916     Gregory Peck   1329962459
1     279      916      need to own   1329962471
2     279      916  romantic comedy   1329962476
3     279      916             Rome   1329962490
4     279      916          royalty   1329962474

In [7]:
map_users = {user: idx for idx, user in enumerate(df.userId.unique())}
map_items = {item: idx for idx, item in enumerate(df.movieId.unique())}
df['userId'] = df['userId'].map(map_users)
df['movieId'] = df['movieId'].map(map_items)
movies_tags['userId'] = movies_tags['userId'].map(map_users)
movies_tags['movieId'] = movies_tags['movieId'].map(map_items)
map_title = {}

for _, row in df.iterrows():
    map_title[row.movieId] = row.title


In [8]:
movies_tags.head()

userId  movieId              tag  timestamp_y
0   11080       34     Gregory Peck   1329962459
1   11080       34      need to own   1329962471
2   11080       34  romantic comedy   1329962476
3   11080       34             Rome   1329962490
4   11080       34          royalty   1329962474

### Divisão da base em treino e teste

In [22]:
from sklearn.model_selection import train_test_split
train, test = train_test_split(df, test_size=.2, random_state=2)

***Exercício 01:*** Aplique a filtragem baseada em conteúdo (ItemAttributeKNN do CaseRecommender) com as tags associadas aos filmes da base. Utilize Jaccard como métrica de similaridade e k=5 vizinhos para predição.

Documentação do ItemAttributeKNN: https://github.com/caserec/CaseRecommender/blob/master/caserec/recommenders/rating_prediction/item_attribute_knn.py

In [20]:
# install caserec
import sys
!{sys.executable} -m pip install caserecommender

In [61]:
# TODO
# Import the necessary library
from caserec.recommenders.rating_prediction.item_attribute_knn import ItemAttributeKNN

train.to_csv('train.dat', index=False, header=False, sep='\t')
test.to_csv('test.dat', index=False, header=False, sep='\t')

# Create a file for the movie tags
movies_tags[['movieId', 'tag']].to_csv('items_tags.dat', index=False, sep='\t', header=False)

# Use ItemAttributeKNN with movie tags as metadata with Jaccard similarity
recommender = ItemAttributeKNN('train.dat', 'test.dat', metadata_file='items_tags.dat',similarity_metric='jaccard', as_similar_first=True, k_neighbors=5)


# Compute recommendations
recommender.compute()


[Case Recommender: Rating Prediction > Item Attribute KNN Algorithm]

train data:: 11090 users and 403 items (152496 interactions) | sparsity:: 96.59%
test data:: 10503 users and 340 items (38125 interactions) | sparsity:: 98.93%

training_time:: 6.768037 sec
>> metadata:: 231 items and 1979 metadata (6274 interactions) | sparsity:: 98.63%
prediction_time:: 0.634239 sec
Eval:: MAE: 0.737907 RMSE: 0.975971 


### Preparação para o exercício 2 - Download e extração de metadados multimídia

In [11]:
!python3 -m wget https://github.com/mmanzato/MBABigData/raw/master/ml-20m-features.tar.gz
! tar -xvzf ml-20m-features.tar.gz


Saved under ml-20m-features.tar.gz
features/
features/._m4infus_max_histogram_300_sn.arq
features/m4infus_max_histogram_300_sn.arq
features/._mm_avg_histogram_100_sn.arq
features/mm_avg_histogram_100_sn.arq
features/._visual_histogram_100_sn.arq
features/visual_histogram_100_sn.arq
features/._visual_histogram_50_sn.arq
features/visual_histogram_50_sn.arq
features/._aural_histogram_50.arq
features/aural_histogram_50.arq
features/._mm_max_histogram_300.arq
features/mm_max_histogram_300.arq
features/._m4infus_max_histogram_50.arq
features/m4infus_max_histogram_50.arq
features/._mm_max_histogram_100.arq
features/mm_max_histogram_100.arq
features/._mm_max_histogram_50_sn.arq
features/mm_max_histogram_50_sn.arq
features/._visual_histogram_100.arq
features/visual_histogram_100.arq
features/._visual_histogram_300.arq
features/visual_histogram_300.arq
features/._aural_histogram_100_sn.arq
features/aural_histogram_100_sn.arq
features/._mm_avg_histogram_100.arq
features/mm_avg_histogram_100.arq


In [30]:
import pickle

with open('./features/visual_histogram_50.arq', 'rb') as arq_visualHistograms:
    visualHistograms = pickle.load(arq_visualHistograms)
print('No. movies: ' + str(len(visualHistograms)))
print('Size of each word: ' + str(len(visualHistograms[0])))

No. movies: 433
Size of each word: 50


In [13]:
visualHistograms[0]

array([0.00948992, 0.00237248, 0.00355872, 0.01304864, 0.05338078,
       0.00711744, 0.05456702, 0.00355872, 0.0059312 , 0.00355872,
       0.0059312 , 0.00474496, 0.00711744, 0.04507711, 0.07591934,
       0.00355872, 0.1316726 , 0.05338078, 0.00830368, 0.        ,
       0.01304864, 0.00474496, 0.02253855, 0.00355872, 0.02728351,
       0.00237248, 0.02016607, 0.00830368, 0.0059312 , 0.00237248,
       0.00355872, 0.04033215, 0.01067616, 0.00237248, 0.00830368,
       0.05931198, 0.0059312 , 0.01067616, 0.02965599, 0.01423488,
       0.00711744, 0.02609727, 0.        , 0.00237248, 0.01423488,
       0.03084223, 0.02016607, 0.00118624, 0.08778173, 0.02253855])

***Exercício 02:*** Como visto, o algoritmo ItemAttributeKNN pode ser usado com diferentes tipos de metadados, como gêneros, tags e palavras no geral. Mais do que isso, podemos adaptá-lo também para que a similaridade entre itens seja feita com base em informações multimídia, como imagens, áudio, etc.

A base de dados utilizada até o momento, ml-20m-compact.tar.gz possui, além das interações de usuários com filmes, uma série de arquivos que contém informações multimídia que foram extraídas dos trailers de cada filme. Esses arquivos estão condensados no zip ml-20m-features.tar.gz, o qual foi feito o download e extraído acima.

Considere por exemplo o arquivo visual_histogram_50.arq. Ele possui 433 vetores (no. de filmes) de tamanho 50. Podemos pensar que cada vetor desse representa informações visuais (cor, brilho, imagem, etc.) que foram extraídas dos trailers de cada filme.

Sua tarefa é usar esses vetores de características visuais no cálculo de similaridade entre os filmes, e em seguida, aplicar essas similaridades no algoritmo ItemAttributeKNN para gerar recomendações.

Dica 1: para calcular a similaridade entre dois vetores pode-se usar o ângulo de cosseno (vide https://en.wikipedia.org/wiki/Cosine_similarity).

Dica 2: é possível passar para o algoritmo ItemAttributeKNN a matriz de similaridade entre itens, por meio do parâmetro similarity_file=arquivo. Veja em: https://github.com/caserec/CaseRecommender/blob/master/caserec/recommenders/rating_prediction/item_attribute_knn.py

In [37]:
# TODO
from sklearn.metrics.pairwise import cosine_similarity
from caserec.recommenders.rating_prediction.item_attribute_knn import ItemAttributeKNN

# Load the visual features
with open('./features/visual_histogram_50.arq', 'rb') as arq_visualHistograms:
    visualHistograms = pickle.load(arq_visualHistograms)

# Calculate cosine similarity between visual features vectors
visual_similarity_matrix = cosine_similarity(visualHistograms)

# Round the similarity matrix values to integers
visual_similarity_matrix_rounded = np.round(visual_similarity_matrix).astype(int)

# Save the rounded similarity matrix to a file with tab as the separator
np.savetxt('visual_similarity_matrix.txt', visual_similarity_matrix_rounded, delimiter='\t', fmt='%d')

# Use ItemAttributeKNN with visual similarity as metadata
recommender = ItemAttributeKNN('train.dat', 'test.dat', similarity_file='visual_similarity_matrix.txt', as_similar_first=True, k_neighbors=5)

# Compute recommendations
recommender.compute()


[Case Recommender: Rating Prediction > Item Attribute KNN Algorithm]

train data:: 11090 users and 403 items (152496 interactions) | sparsity:: 96.59%
test data:: 10503 users and 340 items (38125 interactions) | sparsity:: 98.93%

training_time:: 7.817761 sec
prediction_time:: 0.476343 sec
Eval:: MAE: 0.961725 RMSE: 1.215458 


***Exercício 03:*** Implementar uma função que retorna a probabilidade de um item ser relevante para um usuário, considerando os gêneros dos filmes. Utilize métodos probabilísticos.
- A partir do conjunto de treinamento, obter todas as interações do usuário u.
- Rotular as notas desse usuário como: item relevante se nota >=3 e não relevante se nota < 3.
- Dado um item do conjunto de teste, aplicar o Teorema de Bayes com suavização de Laplace. Utilizar os gêneros associados.
- Retornar se o item é ou não relevante, e em seguida, comparar o resultado com a nota real que esse usuário deu para o item (disponível no conjunto de teste).

In [66]:
# TODO
import pandas as pd
import numpy as np

# Build movies_genres
movies_genres = movies.drop('genres', axis=1).join(movies.genres.str.split('|', expand=True)
             .stack().reset_index(drop=True, level=1).rename('genre'))
movies_genres.dropna(inplace=True)
movies_genres['movieId'] = movies_genres.movieId.astype(int)

# Extract unique genres from the 'movies_genres' DataFrame
genres_unique = movies_genres['genre'].unique()

# Step 1: Calculate relevance probabilities for each genre based on training ratings
def calculate_genre_probabilities(train):
    genre_probabilities = {}
    total_ratings = len(train)

    for genre in genres_unique:
        genre_ratings = train[train['movieId'].isin(movies_genres[movies_genres['genre'] == genre]['movieId'])]
        relevant_ratings = genre_ratings[genre_ratings['rating'] >= 3]
        probability = (len(relevant_ratings) + 1) / (len(genre_ratings) + 2)  # Laplace smoothing
        genre_probabilities[genre] = probability

    return genre_probabilities

# Step 2: Calculate the probability of an item being relevant to a user based on the item's genres
def calculate_item_probability(user_id, item_genres, genre_probabilities):
    user_ratings = train[train['userId'] == user_id]
    user_relevant_genres = set()

    for index, row in user_ratings.iterrows():
        movie_genres = movies_genres[movies_genres['movieId'] == row['movieId']]['genre'].tolist()
        user_relevant_genres.update(movie_genres)

    item_probability = 1.0

    for genre in item_genres:
        if genre in user_relevant_genres:
            item_probability *= genre_probabilities[genre]
        else:
            item_probability *= (1 - genre_probabilities[genre])

    return item_probability

# Step 3: Compare the calculated probability with the user's actual rating in the test set
def evaluate_relevance_predictions(test, genre_probabilities):
    correct_predictions = 0
    total_predictions = len(test)

    for index, row in test.iterrows():
        user_id = row['userId']
        item_genres = movies_genres[movies_genres['movieId'] == row['movieId']]['genre'].tolist()
        item_probability = calculate_item_probability(user_id, item_genres, genre_probabilities)

        if item_probability >= 0.5 and row['rating'] >= 3:
            correct_predictions += 1
        elif item_probability < 0.5 and row['rating'] < 3:
            correct_predictions += 1

    accuracy = correct_predictions / total_predictions
    return accuracy


# Calculate the relevance probabilities for each genre based on training data
genre_probabilities = calculate_genre_probabilities(train)

# Evaluate the accuracy of relevance predictions on the test set
accuracy = evaluate_relevance_predictions(test, genre_probabilities)
print(f'The accuracy of relevance predictions is {accuracy * 100:.2f}%')


The accuracy of relevance predictions is 76.78%
